### PR TITLE
api: projects: fetch user if missing subitem

### PR DESCRIPTION
**Pull request by gioelecerati**
**What does this pull request do? Explain your changes. (required)**

When the document is missing the subitem `user`, fetch it based on the document.userId

**Specific updates (required)**

**How did you test each of these updates (required)**

**Does this pull request close any open issues?**

**Screenshots (optional)**

**Checklist**

- [ ] I have read the **CONTRIBUTING** document.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.

### CODE REVIEW

**Review by leszko — APPROVED** [review]
LGTM. It'd be nice if you could add a unit test.

File: packages/api/src/controllers/helpers.ts [hunk 1]
```diff
@@ -771,45 +771,49 @@ export async function addDefaultProjectId(
...
 
-  const enrichResponseWithUserProjectId = (document) => {
+  const enrichResponseWithUserProjectId = async (document) => {
     if ("id" in document && "userId" in document && "user" in document) {
-      if (
-        (!document.projectId || document.projectId === "") &&
-        document.user?.defaultProjectId
-      ) {
-        document.projectId = document.user.defaultProjectId;
+      if (!document.projectId || document.projectId === "") {
```
Comment (victorges) [inline packages/api/src/controllers/helpers.ts]: **[js-wat](https://www.destroyallsoftware.com/talks/wat)-protip**: empty string is already falsy! same behavior:
```suggestion
      if (!document.projectId) {
```

File: packages/api/src/controllers/helpers.ts [hunk 2]
```diff
@@ -771,45 +771,49 @@ export async function addDefaultProjectId(
...
+        if (document.user?.defaultProjectId) {
+          document.projectId = document.user.defaultProjectId;
```
Comment (victorges) [inline packages/api/src/controllers/helpers.ts]: We shouldn't be fetching the user from the DB again if the user is already embedded in the object.
```suggestion
        if (document.user) {
          document.projectId = document.user.defaultProjectId;
```

File: packages/api/src/controllers/helpers.ts [hunk 3]
```diff
@@ -771,45 +771,49 @@ export async function addDefaultProjectId(
     }
   };
 
-  const enrichResponseWithUserProjectId = (document) => {
+  const enrichResponseWithUserProjectId = async (document) => {
     if ("id" in document && "userId" in document && "user" in document) {
```
Comment (victorges) [inline packages/api/src/controllers/helpers.ts]: This condition needs to be updated.

File: packages/api/src/controllers/helpers.ts [hunk 4]
```diff
@@ -771,45 +771,49 @@ export async function addDefaultProjectId(
...
-      if (
-        (!document.projectId || document.projectId === "") &&
-        document.user?.defaultProjectId
-      ) {
-        document.projectId = document.user.defaultProjectId;
+      if (!document.projectId || document.projectId === "") {
+        if (document.user?.defaultProjectId) {
+          document.projectId = document.user.defaultProjectId;
+        } else {
+          let user = await db.user.get(document.userId, { useCache: true });
```
Comment (victorges) [inline packages/api/src/controllers/helpers.ts]: We shouldn't be fetching the user if they are already the same as the caller.
```suggestion
          let user = document.userId === req.user.id ? req.user : await db.user.get(document.userId, { useCache: true });
```
(good call on caching)

File: packages/api/src/controllers/helpers.ts [hunk 5]
```diff
@@ -771,45 +771,49 @@ export async function addDefaultProjectId(
...
+          if (user.defaultProjectId) {
+            document.projectId = user.defaultProjectId;
+          }
```
Comment (victorges) [inline packages/api/src/controllers/helpers.ts:786]: It's not really a problem if `defaultProjectId` is not set, right? We'd just be setting the `projectId` to empty or undefined, which it actually is anyway (if it has anything else it's a mistake!)

Not a problem checking it either, just feels like unnecessary branches to be reasoning about in the code, apart from the required ones.

```suggestion
          document.projectId = user.defaultProjectId;
```

File: packages/api/src/controllers/helpers.ts [hunk 6]
```diff
@@ -771,45 +771,49 @@ export async function addDefaultProjectId(
...
 
-      Object.values(item).forEach((subItem) => {
-        if (typeof subItem === "object" && subItem !== null) {
-          if (req.user.admin) {
-            enrichResponseWithUserProjectId(subItem);
-          } else {
-            enrichResponse(subItem);
+      await Promise.all(
+        Object.values(item).map(async (subItem) => {
+          if (typeof subItem === "object" && subItem !== null) {
```
Comment (victorges) [inline packages/api/src/controllers/helpers.ts:804]: Remind me why do we need this recursion again? Do we have project-objects nested inside project-objects as well?

File: packages/api/src/controllers/helpers.ts (hunk 3 above)
Comment (gioelecerati) [reply packages/api/src/controllers/helpers.ts]: Am I tripping? I already see the `async` in the changes

File: packages/api/src/controllers/helpers.ts (hunk 6 above)
Comment (gioelecerati) [reply packages/api/src/controllers/helpers.ts:804]: Wasn't it for the cases like the asset creation, which returns {asset:{},task:{}}?

File: packages/api/src/controllers/helpers.ts (hunk 3 above)
Comment (gioelecerati) [reply packages/api/src/controllers/helpers.ts]: Ah I see what you're referring to, fixed it

File: packages/api/src/controllers/helpers.ts (hunk 6 above)
Comment (victorges) [reply packages/api/src/controllers/helpers.ts:804]: Ahh yeah right!

**Review by victorges — APPROVED** [review]

**Pull request merged**